### PR TITLE
Support POSIX AIO in mio again

### PR DESCRIPTION
**Pull request by asomers**
* impl AsRawFd for Registry
* Add smoke tests for POSIX AIO

See also https://github.com/asomers/mio-aio

### CODE REVIEW

**Comment by asomers** [comment]
The OSX test failures look unrelated.

File: tests/aio.rs [hunk 1]
```diff
@@ -0,0 +1,226 @@
...
+
+mod aio {
+    use super::*;
+
+    #[test]
+    fn smoke() {
+        let mut poll = Poll::new().unwrap();
+        let mut events = Events::with_capacity(8);
+
+        let f = tempfile().unwrap();
```
Comment (Thomasdezeeuw) [inline tests/aio.rs:183]: Nit: can you replace this with `temp_file("aio::smoke")` from the `test::util` module? Then we don't need the additional dependency.

File: tests/aio.rs [hunk 2]
```diff
@@ -0,0 +1,226 @@
...
+    use super::*;
+
+    #[test]
+    fn smoke() {
+        let data = b"hello, world!\n";
+        let mut poll = Poll::new().unwrap();
+        let mut events = Events::with_capacity(8);
+
+        let f0 = tempfile().unwrap();
+        let f1 = tempfile().unwrap();
```
Comment (Thomasdezeeuw) [inline tests/aio.rs:210]: Nit: same as above, can you replace this with `temp_file("lio::smoke0/1")` from the `test::util` module?

File: tests/aio.rs [hunk 3]
```diff
@@ -0,0 +1,226 @@
...
+    pub fn from_fd(fd: RawFd) -> Aiocb {
+        // Use mem::zeroed instead of explicitly zeroing each field, because the
+        // number and name of reserved fields is OS-dependent.  On some OSes,
+        // some reserved fields are used the kernel for state, and must be
+        // explicitly zeroed when allocated.
+        let mut inner = unsafe { mem::zeroed::<libc::aiocb>() };
+        inner.aio_fildes = fd;
+        inner.aio_sigevent.sigev_notify = libc::SIGEV_NONE;
+        inner.aio_offset = 0;
+        inner.aio_nbytes = 0;
```
Comment (Thomasdezeeuw) [inline tests/aio.rs:36]: These aren't needed with zeroing above right?

File: tests/aio.rs [hunk 4]
```diff
@@ -0,0 +1,226 @@
...
+        aiocb
+    }
+
+    pub fn fsync(&mut self) -> Result<(), i32> {
+        unsafe {
+            // Safe because we don't move the libc::aiocb
+            let selfp = self.0.as_mut().get_unchecked_mut();
+            match libc::aio_fsync(libc::O_SYNC, selfp) {
+                0 => Ok(()),
+                _ => Err(*libc::__error()),
```
Comment (Thomasdezeeuw) [inline tests/aio.rs:54]: Nit: why not use `io::Error::last_os_error` and return an `io::Error`?

File: tests/aio.rs [hunk 5]
```diff
@@ -0,0 +1,226 @@
...
+        unsafe {
+            let r = libc::lio_listio(
+                libc::LIO_NOWAIT,
+                self.list.as_ptr(),
+                self.list.len() as i32,
+                &mut self.sev as *mut libc::sigevent,
+            );
+            match r {
+                0 => Ok(()),
+                _ => Err(*libc::__error()),
```
Comment (Thomasdezeeuw) [inline tests/aio.rs:117]: Nit: same as above, why not use `io::Error::last_os_error` and return an `io::Error`?

**Review by Thomasdezeeuw — APPROVED** [review]
Add some comments about some small things, but overall LGTM.

**Comment by Thomasdezeeuw** [comment]
I forgot you need to call `init` from the test module before using `temp_file`. If you can fix that and squash the last two commits (leaving 2 commits; `AsRawFd` and tests), I'm good with merging.

**Comment by Thomasdezeeuw** [comment]
Thanks @asomers.

**Pull request merged**